File: content/snippets/detect-spam-in-contact-form/n1.py
"""
Tell spam from a real enquiry with a linear classifier on character n-grams.

Rung N1. The rules of N0 look for the words on a list. This one scores
the character fragments of a message, with weights learnt from labelled
submissions, taken from those the inbox has already received.

Character n-grams rather than words, so `backlink`, `backlinks` and
`backlinking` share most of their features. A word spelt out letter by letter,
`b a c k l i n k s`, shares none: no n-gram crosses a space.

The decision is a weighted sum, so you can print the features that pushed a
message over the line, which matters the first time someone asks why their
enquiry was rejected.
"""

import unicodedata

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline


def fold(text: str) -> str:
    """Lowercase and strip accents, so casing never doubles the feature space."""
    stripped = unicodedata.normalize("NFKD", text)
    return "".join(c for c in stripped if not unicodedata.combining(c)).casefold()


def train(messages: list[str], labels: list[int]):
    """`labels` is 1 when the submission is spam, 0 when it is a real enquiry."""
    model = make_pipeline(
        # `char_wb` keeps n-grams inside word boundaries, so the model learns
        # word shapes rather than the way two neighbours happen to collide.
        TfidfVectorizer(analyzer="char_wb", ngram_range=(3, 5), sublinear_tf=True, min_df=1),
        # Balanced, so whichever class is rarer in your labels weighs as much
        # as the other. `C` above one because, on a small labelled set, a
        # heavy regulariser leaves every score sitting near a half, which
        # makes the threshold below meaningless.
        LogisticRegression(class_weight="balanced", C=10.0, max_iter=1000),
    )
    model.fit([fold(m) for m in messages], labels)
    return model


def spam_score(model, message: str) -> float:
    """Probability that the submission is spam, between zero and one."""
    folded = fold(message)
    # Nothing to read: without this, the model's bias alone would decide.
    if not folded.strip():
        return 0.0
    return float(model.predict_proba([folded])[0][1])


def is_spam(model, message: str, threshold: float = 0.5) -> bool:
    """
    Returns a decision, and the threshold is yours to set.

    Move it towards 1 when losing a real enquiry is the expensive mistake.
    Move it towards 0 when a spam message reaching a human is the expensive one.
    """
    return spam_score(model, message) >= threshold

File: content/snippets/detect-spam-in-contact-form/n1.js
/**
 * Tell spam from a real enquiry with a linear classifier on character n-grams.
 *
 * Rung N1. The rules of N0 look for the words a spammer used last year. This
 * looks at how the message is written: a few hundred labelled submissions, the
 * kind an inbox already holds, and the model learns the register rather than
 * the vocabulary list.
 *
 * Written out in full rather than pulled from a library, because TF-IDF on
 * character n-grams and a logistic regression fit by gradient descent is forty
 * lines. That is the whole argument of this rung: the classical tool is small
 * enough to read.
 */

const BUCKETS = 1024; // hashing trick: no vocabulary to build, or to ship
const NGRAMS = [3, 4, 5]; // long enough to carry a word, short enough to survive a typo

/** Lowercase and strip accents, so casing never doubles the feature space. */
export function fold(text) {
  return text.normalize('NFKD').replace(/\p{Diacritic}/gu, '').toLowerCase();
}

/** How often each hashed character n-gram occurs in the message. */
function counts(text) {
  const padded = ` ${fold(text)} `;
  const seen = new Float64Array(BUCKETS);
  for (const n of NGRAMS) {
    for (let i = 0; i + n <= padded.length; i += 1) {
      let h = 2166136261;
      for (const c of padded.slice(i, i + n)) h = ((h ^ c.codePointAt(0)) * 16777619) >>> 0;
      seen[h % BUCKETS] += 1;
    }
  }
  return seen;
}

/** Sublinear term frequency, weighted by inverse document frequency, L2 normalised. */
function vector(seen, idf) {
  const v = seen.map((count, j) => (count ? (1 + Math.log(count)) * idf[j] : 0));
  const norm = Math.sqrt(v.reduce((sum, x) => sum + x * x, 0));
  return norm ? v.map((x) => x / norm) : v;
}

/** `labels` is 1 when the submission is spam, 0 when it is a real enquiry. */
export function train(messages, labels, { epochs = 300, rate = 0.5 } = {}) {
  const raw = messages.map(counts);
  // Inverse document frequency: an n-gram every message carries says nothing.
  const idf = new Float64Array(BUCKETS).map((_, j) =>
    Math.log((1 + raw.length) / (1 + raw.filter((seen) => seen[j] > 0).length)) + 1);
  const rows = raw.map((seen) => vector(seen, idf));
  const model = { weights: new Float64Array(BUCKETS), bias: 0, idf };

  for (let epoch = 0; epoch < epochs; epoch += 1) {
    for (let i = 0; i < rows.length; i += 1) {
      const error = probability(model, rows[i]) - labels[i];
      for (let j = 0; j < BUCKETS; j += 1) model.weights[j] -= rate * error * rows[i][j];
      model.bias -= rate * error;
    }
  }
  return model;
}

/** The logistic of the weighted sum: one number between zero and one. */
function probability(model, row) {
  let z = model.bias;
  for (let j = 0; j < BUCKETS; j += 1) z += model.weights[j] * row[j];
  return 1 / (1 + Math.exp(-z));
}

/** Probability that the submission is spam. */
export function spamScore(model, message) {
  return probability(model, vector(counts(message), model.idf));
}

/**
 * Returns a decision, and the threshold is yours to set.
 *
 * Move it towards 1 when losing a real enquiry is the expensive mistake.
 * Move it towards 0 when a spam message reaching a human is the expensive one.
 */
export function isSpam(model, message, threshold = 0.5) {
  return spamScore(model, message) >= threshold;
}
